### ISSUE TITLE
Unity Fractal Server + Unity Hololens Client connect, but no video

### ISSUE THREAD

**Issue by treisto** [CatalystCode/3dtoolkit#187]
Hello,

I have installed the latest master branch, on the server laptop I have Nvidia GTX950M, CUDA development tools installed, NVIDIA driver reports 397.31, I have VS 2015.3, unity 5.6.3.f1, win 10; ran the ./setup.cmd successfully. Built the server in x64 and the client in x86. Added the server exec to run on NVIDIA card from NVIDIA options software (I have dual card laptop).  

For signaling, I run a local wi-fi from a dedicated Linksys consumer router, with no internet access.
The signalling server is the one from anastasiia-zolochevska/signaling-server.

The client and server connect, but on the server I still have the Connect UI on black background, and on the Hololens I have just a black screen. The client log reads "no decodable frame". I have set the server to 30 fps (default is on 60), as I see the client should be on 30. 

The server sends frames successfully to the dx-win32 client at 60 fps, but even in this configuration, the UI+black screen still remains on the server.

Please find attached the log files and config files for server and client.
Any help would be very much appreciated.
[LOGS.ZIP](https://github.com/CatalystCode/3dtoolkit/files/1988027/LOGS.ZIP)

**Comment by phongcao** [comment]
@treisto Thanks for the logs. Looks like there might be some issues with WebRTC config file. Can you try changing the port of your signaling server to **80** and modify your config file like this:

`
{
  "iceConfiguration": "none",
  "server": "http://192.168.1.2",
  "port": 80,
  "heartbeat": 5000
}
`

**Comment by treisto** [comment]
Hello and thank you very much for the help. I have tried changing the signalling port to 80, and webrtc config files in client and server, but the result is still the same. Please find attached the new logs.
I'm gonna also try and reset the router to factory settings; and will also try to find another router for test.
[LOGS_1.ZIP](https://github.com/CatalystCode/3dtoolkit/files/1988518/LOGS_1.ZIP)

**Comment by treisto** [comment]
Update: resetting the router didn't help. 
I also quickly created a wi-fi hotspot with the phone. 
In both cases client and server connected with same behavior.

**Comment by phongcao** [comment]
@treisto It does work if you use SpinningCube/Multithreaded server with the HoloLens client correct?

**Comment by phongcao** [comment]
And you're right that internet is not mandatory on local wi-fi.

**Comment by treisto** [comment]
I am not 100% sure but I think not - those are also only connecting and then, white screens - I think it was white screen on those servers. Unfortunately I left (already evening where I am) and I have no access to Hololens until Friday. Friday I will do more tests and will update here.

**Comment by treisto** [comment]
Indeed I was thinking on smthng related to stereo, because UnityFractalServer -> DxWin32 (mono) works. I will also try 50.000 on Friday. Thank you very much.

**Comment by phongcao** [comment]
In that case, I'd love to see the logs that you can get from the DirectX server samples (using **Debug** configuration) since it provide much more helpful info for us to debug.

**Comment by machur** [comment]
@treisto From my experience I'd recommend you setting up leightweight SpinningCubeServer and DirectX client for HL first. I had similar problem with no video visible that was related to network capacity. It seems that significant time delays cause issues with prediction timestamps that may result in "no decodable frame" messages.

Also make sure that you start client application while having HL pointed in proper direction to avoid displaying objects in some weird location e.g. behind you - it may sound silly, but it happens :wink:

**Comment by treisto** [comment]
@phongcao @machur Thank you both very much for the hints. 

In the end I was able to connect and render some image on all combinations Win32/DxHolo/UnityHolo clients <-> Multithread/Spinning Cube/Fractal Servers - more or less correctly, and with various problems or performance issues, but they worked.

Except for the pair UnityHoloClient <-> UnityFractalServer which still does not work, even after I set  StereoFlagWaitTime in WebRTCServer.cs in Unity Server to 50.000 - unfortunately exactly the pair of interest to me. I tried release and debug versions, with and without debugger attached. I will investigate further on Monday, but any new suggestions on this in the meanwhile would be very welcome.

For all servers - the Connecting UI is not disappearing after connecting to client. 
I saw in some other screenshots here that the rendered image is also displayed on the server side?

For the UnityHololensClient - there are no stats on-screen implemented yet, or should I also check why I don't have this working?

![server1](https://user-images.githubusercontent.com/4274128/39933237-afcefab4-5542-11e8-9094-2fd659df1a0c.jpg)

==============

Somewhat off-topic, if you'd allow me to give one feedback - the text displayed on the white screen of the servers "Multithread" and "Spinning Cube" just after connecting to the signalling server, and before selecting a peer for starting streaming to it seems a bit non-intuitive in my opinion. The text displays "List of currently connected peers:" which can make one think that everything is fine and no more action is needed. So I was wondering why nothing happens further. In fact one must select the desired peer and press "Enter" to start streaming. It took me a while to figure this out :D so maybe it would be useful for other people in the future if the text would be more descriptive like maybe : "List of peers currently connected to the Signalling Server. Please select the desired peer then press "Enter" to start streaming to it". 

Thank you very much once again and a nice weekend

**Comment by treisto** [comment]
I have checked the wi-fi channels in my area. I was on 6, I moved and I am now on 9 - the left orange "00:1E:E5..."
The UnityFractal Server is connected to the router by cable, so only the Hololens is on Wi-Fi. 
The router is within 1meter from Hololens. 

Can it be that bad? Should I buy a 5Ghz router - I see the 5Ghz is less crowded? (mine is only 2.4Ghz)

<img width="1277" alt="screen shot 2018-05-14 at 14 51 49" src="https://user-images.githubusercontent.com/4274128/39999044-34c6d5ba-5788-11e8-8a20-111213a23c4f.png">

**Comment by winodesv** [comment]
@treisto  Except for the pair UnityHoloClient <-> UnityFractalServer which still does not work,

Did you mean, UnityHoloClient is working with everything but UnityFractalServer or UnityFractalServer is working with everything but not with HoloClient?

**Comment by treisto** [comment]
@winodesv 

- UnityHoloClient works with the SpinningCubeServer (although I see the cube only sometimes in weird positions and it looks like it is moving)
- UnityHoloClient works with the MultithreadServer (low frame and no stats on-screen)

- UnityFractalServer works with DxWin32 client (60fps)
- UnityFractalServer works with DxHoloClient 

- UnityHoloClient and UnityFractalServer connect, but I don't get any video (in the client, on debug, I see "no decodable frame")

**Comment by winodesv** [comment]
@phongcao UnityFractalServer crashing with streamingclient. Can you guide me where I went wrong?
[error.log](https://github.com/CatalystCode/3dtoolkit/files/2004527/error.log)

@treisto Thanks for details.

**Comment by winodesv** [comment]
@phongcao Oh forgot to add unity log. Here it is...
[output_log.txt](https://github.com/CatalystCode/3dtoolkit/files/2004765/output_log.txt)

**Comment by treisto** [comment]
@winodesv

Note: Unity-FractalStream requires Unity 5.6.3f1 or higher (we test at 5.6.3f1)

**Comment by winodesv** [comment]
@treisto tried with unity 2017.3.0 and 2017.4.2 . Crashed with both. 
Will try 5.6.3f1.

**Comment by phongcao** [comment]
@winodesv Did you run the FractalStream server with [-force-d3d11-no-singlethreaded](https://github.com/CatalystCode/3dtoolkit/tree/master/Samples/Server/Unity-FractalStream) flag?

**Comment by phongcao** [comment]
@treisto The Unity Fractal Server + Unity Hololens Client pair doesn't work at the moment. We'll try to have a look at it as soon as we can and let you know. Regarding your concern about wifi router, @anderm should be able to provide more insights. In the coming weeks, I also plan to release some performance improvements for DirectX HoloLens client, which will improve the network latency a bit more.

**Comment by winodesv** [comment]
Oops Phongcao, you are right. I forgot the flag. It works now.

**Comment by treisto** [comment]
@phongcao Thank you very much, at least I can stop trying for now. In the mean time I will continue with the DxHoloClient. If you could have a look at the UnityClient when you have a chance, it would be of very great help.

**Comment by treisto** [comment]
@phongcao could you please tell me what is the meaning of fps displayed on DxHoloClient? Is it the true final fps displayed on Holo screen?

**Comment by treisto** [comment]
@anderm Hi Andrei, could you please advise what would be the recommended router for the Unity Fractal Server to work ok with the Dx and Unity Hololens Clients? 

Right now, with the 2.4Ghz b/g router that I have, the DxHolo Client works with spikes in frame/latency (2 - 30 fps), and the Unity Holo Client doesn't work at all (I get "no decodable frame"). The router is used exclusively for this, and is on my desk within 1 meter from the HoloLens. The server is a laptop with Nvidia GTX950M, connected by cable to the router. 

Should I upgrade to a 5Ghz a/c router? 
Please also see the screenshot in the thread, above.

Thank you very much,
Lucian

**Comment by treisto** [comment]
@winodesv Hi, does your Unity Fractal server work with Unity 2017, or Unity 5.6, or both?

**Comment by winodesv** [comment]
@treisto Its working with both.

**Comment by winodesv** [comment]
My next trouble is making stereo work. 

@phongcao Currently I don't have access to a Hololens, but I have a Acer WMR device with me. What is the best server-client combination to test on WMR device? Or is it better to try on Hololens emulator?

**Comment by phongcao** [comment]
@winodesv I'm afraid that you would need a physical HoloLens device to test since it doesn't work with Acer WMR device and Hololens emulator doesn't support H264 codec.

**Comment by treisto** [comment]
@phongcao Hi phongcao, when using UnityFractal Server with Streaming Dx client, the image is upside down. I see there is this shader and script FlipY plus a material matRT, what do I have to do with them, please?

**Comment by phongcao** [comment]
@treisto You need to enable **UNITY_UV_STARTS_AT_TOP** flag in the client. Check it [here](https://github.com/CatalystCode/3dtoolkit/tree/master/Samples/Server/Unity-FractalStream).

**Comment by treisto** [comment]
@phongcao I had a feeling I saw something before but I just couldn't find this anymore. Sorry for having bothered you for something that was already there. Thanks a lot

**Comment by treisto** [comment]
Hello all, do you think it could be any chance that this would be fixed in the near future - June, July, please?
It would add tremendous value to our (3D medical visualisations) project as some features would be best implemented in the Unity Holo client.

**Comment by phongcao** [comment]
@treisto Even if we can make the Unity Holo client work with the Unity FractalStream server, it still doesn't have the image stabilization to keep the hologram **world-locked**. Would you still be okay with that?

**Comment by anderm** [comment]
@treisto That is great to know. Unfortunately, at the moment, we do not have any Unity work scheduled until end-June but we are planning to look at improving our HoloLens server and Unity client around August time-frame with a few of our partners. Would that work for your current efforts? It would also be very valuable if we can get an idea of what you are trying to develop, your current customers and when you plan to go to market. Can you please send me an email at `anderm@microsoft.com` with the details? We can then look at prioritizing our work for the upcoming period. 

Thank you for your interest in our toolkit!

**Comment by treisto** [comment]
For anybody using the HoloLens: Just wanna share that, at least for us, by using a 5Ghz router, the pair Dx HoloLens  Client + Unity Fractal Server felt a bit more snappy and responsive, with somewhat less hiccups (vs a 2.4 Ghz router). While the framerate is the same, latency seems to do a little bit better.

**Comment by iamrohit1** [comment]
If anyone stumbling here trying to get unity server and client to work:

I was able to get the default server and client (both unity) to connect and stream to each other without any code modification. Only issue was client was receiving random frames.

Problem was the rotation component was jumping all over the place. 
Here's an easy fix on server : Change 
```
Quaternion leftQ = QuaternionFromMatrix(peerData.stereoLeftViewMatrix);
this.LeftEye.transform.rotation = new Quaternion(-leftQ.x, -leftQ.y, leftQ.z, leftQ.w);
```
on this [line](https://github.com/3DStreamingToolkit/3DStreamingToolkit/blob/master/Samples/Server/Unity-FractalStream/Assets/Scripts/WebRTCServer.cs#L218) to
```
Quaternion leftQ = QuaternionFromMatrix(peerData.stereoLeftViewMatrix * Matrix4x4.TRS(Vector3.zero, Quaternion.identity, new Vector3(1, 1, -1)));
this.LeftEye.transform.rotation = new Quaternion(leftQ.x, leftQ.y, leftQ.z, leftQ.w);
```
Similarly for the right eye as well. 

Also you'll need to attach the FlipY script to one of the cameras 
if you're using a 3rd camera for UI stuff change 

` Graphics.Blit(src, dest, matRT);`

at [this line](https://github.com/3DStreamingToolkit/3DStreamingToolkit/blob/63356fd4839bedeb39200fe7fc50dc01f333321c/Samples/Server/Unity-FractalStream/Assets/Scripts/FlipY.cs#L13) to 
```
if(src) 
     Graphics.Blit(src, dest, matRT);
```

Hope this helps someone!

**Comment by iamrohit1** [comment]
Another issue with unity client is the matrices returned by left and right eyes are same. As a result you'll see both left and right images same, hence breaking the stereo.

Here are some references:
https://forum.unity.com/threads/world-positions-of-left-right-eye.350076/

Unity's single pass stereo doesnt work either btw:
https://forum.unity.com/threads/is-single-pass-stereo-rendering-supported-for-rendertextures.485294/

I did a dirty fix and handled it on server side offsetting the cameras.

**Comment by iamrohit1** [comment]
I encountered a new issue. With 720p stereo the Unity server and Unity client work well. But for any other resolution the output is blank. 

I accessed the texture the output is being rendered on and made a screenshot of it. Heres the screen cap for `1280*2 x 720`:
![sample](https://user-images.githubusercontent.com/13789318/58714832-742f0c80-837a-11e9-9819-13959d52021f.png)
And heres screenshot for `648*2x360`:
![sample_360](https://user-images.githubusercontent.com/13789318/58714875-8f018100-837a-11e9-9396-f266495daee5.png)
(There is an image above.)
I think whatever is causing this is not on Unity but in the plugin (probably MediaEngineUWP). I've tested with couple of other resolutions as well and had the same blank output. 
(And yes I changed the resolutions correctly on server and the client.)
Just wanted to check if anybody else is working on it and had a solution. I'll update if I find one myself.
Thanks!

**Comment by treisto** [comment]
Hi iamrohit1,

Thanks a lot for the updates.
Are you working on Hololens?

**Comment by iamrohit1** [comment]
Hey,
No Problem. Yes, I'm working on Holo.

**Comment by chansk131** [comment]
> line

How about line 207 and 212 where it updates transform position for each eye. Do we need to change from 

                            this.LeftEye.transform.position = new Vector3(
                                peerData.stereoLeftViewMatrix.m03,
                                peerData.stereoLeftViewMatrix.m13,
                                -peerData.stereoLeftViewMatrix.m23);

                            this.RightEye.transform.position = new Vector3(
                                peerData.stereoRightViewMatrix.m03,
                                peerData.stereoRightViewMatrix.m13,
                                -peerData.stereoRightViewMatrix.m23);`

to 


                            this.LeftEye.transform.position = new Vector3(
                                peerData.stereoLeftViewMatrix.m03,
                                peerData.stereoLeftViewMatrix.m13,
                                peerData.stereoLeftViewMatrix.m23);

                            this.RightEye.transform.position = new Vector3(
                                peerData.stereoRightViewMatrix.m03,
                                peerData.stereoRightViewMatrix.m13,
                                peerData.stereoRightViewMatrix.m23);`

basically delete minus sign of the z axis? Because with the current code, it looks like it is getting closer when we are moving away from the hologram.

**Comment by iamrohit1** [comment]
@chansk131 that should not be necessary if you're using Unity Client (or DirectX I believe) because both are in left hand coordinate system.. (same as Unity server) 

Please check if you're inverting the view matrix before using it.. ([this line right here](https://github.com/3DStreamingToolkit/3DStreamingToolkit/blob/63356fd4839bedeb39200fe7fc50dc01f333321c/Samples/Server/Unity-FractalStream/Assets/Scripts/WebRTCServer.cs#L203))

If you're doubtful that your matrices are not coming out right, you can verify it easily by printing the inverse of view matrix (worldToCameraMatrix) of your camera. (m03,m13,-m23) will give you the position of your camera.
